tools/data_collector_tool/smartapi/data_downloader_utils.py

```diff
@@ -0,0 +1,116 @@
+import time
+from datetime import datetime, timedelta
+from pathlib import Path
+
+import pandas as pd
+import requests
+
+from app.utils.common.types.reques_types import CandlestickInterval
+from app.utils.file_utils import load_json_data, write_to_json_file
+
+
+def search_valid_date(
+    start_date: datetime,
+    end_date: datetime,
+    stock_symbol: str,
+    interval: CandlestickInterval,
+) -> datetime:
+    """It finds the valid month from where the availability of data starts for
+    the given stock symbol and interval by using binary search method.
+
+    Parameters:
+    -----------
+    start_date: ``datetime``
+        Start date to search.
+    end_date: ``datetime``
+        End date to search.
+    stock_symbol: ``str``
+        The symbol of the stock.
+
+    Return:
+    -------
+    datetime
+        searched month from where the availability of data starts for the given stock symbol and interval.
+    """
+    valid_date = end_date
+    while start_date <= end_date:
+        try:
+            total_days = (end_date - start_date).days
+            middle_date = start_date + timedelta(days=total_days // 2)
+            first_day = middle_date.replace(day=3)
+            last_day = (middle_date + timedelta(days=31)).replace(day=1) - timedelta(
+                days=1
+            )
+            stocks_url = (
+                f"http://127.0.0.1:8000/smart-api/equity/history/{stock_symbol}?interval={interval.name}"
+                f"&start_date={first_day.strftime('%Y-%m-%d')}%2009%3A15&end_date={last_day.strftime('%Y-%m-%d')}%2015%3A29"
+            )
```

Comment:
Why are you using stocks_url in two places, i.e data_download.py and data_downloader_utils.py. This is redundant. move these to one place

Reply:
added into the constants variable.